File: memind-clients/rust/src/models/message.rs
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use serde::de;
use serde::ser::SerializeMap;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use time::OffsetDateTime;

use super::Role;
use crate::{MemindError, Result};

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Message {
    pub role: Role,
    pub content: Vec<ContentBlock>,
    #[serde(
        default,
        with = "time::serde::rfc3339::option",
        skip_serializing_if = "Option::is_none"
    )]
    pub timestamp: Option<OffsetDateTime>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_client: Option<String>,
}

impl Message {
    pub fn user(text: impl Into<String>) -> Self {
        Self {
            role: Role::User,
            content: vec![ContentBlock::text(text)],
            timestamp: None,
            user_name: None,
            source_client: None,
        }
    }

    pub fn assistant(text: impl Into<String>) -> Self {
        Self {
            role: Role::Assistant,
            content: vec![ContentBlock::text(text)],
            timestamp: None,
            user_name: None,
            source_client: None,
        }
    }

    pub fn with_timestamp(mut self, timestamp: OffsetDateTime) -> Self {
        self.timestamp = Some(timestamp);
        self
    }

    pub fn with_user_name(mut self, user_name: impl Into<String>) -> Self {
        let trimmed = user_name.into().trim().to_string();
        self.user_name = (!trimmed.is_empty()).then_some(trimmed);
        self
    }

    pub fn with_source_client(mut self, source_client: impl Into<String>) -> Self {
        let trimmed = source_client.into().trim().to_string();
        self.source_client = (!trimmed.is_empty()).then_some(trimmed);
        self
    }

    pub(crate) fn validate(&self) -> Result<()> {
        if self.content.is_empty() {
            return Err(MemindError::invalid_request(
                "message.content must contain at least one block",
            ));
        }
        for block in &self.content {
            block.validate()?;
        }
        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum ContentBlock {
    Text { text: String },
    Image { source: Source },
    Audio { source: Source },
    Video { source: Source },
}

impl ContentBlock {
    pub fn text(text: impl Into<String>) -> Self {
        Self::Text { text: text.into() }
    }

    pub fn image(source: Source) -> Self {
        Self::Image { source }
    }

    pub fn audio(source: Source) -> Self {
        Self::Audio { source }
    }

    pub fn video(source: Source) -> Self {
        Self::Video { source }
    }

    pub(crate) fn validate(&self) -> Result<()> {
        match self {
            Self::Text { .. } => Ok(()),
            Self::Image { source } | Self::Audio { source } | Self::Video { source } => {
                source.validate()
            }
        }
    }
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum Source {
    Url {
        url: String,
    },
    Base64 {
        #[serde(rename = "media_type")]
        media_type: String,
        data: String,
    },
}

impl Source {
    pub fn url(url: impl Into<String>) -> Self {
        Self::Url { url: url.into() }
    }

    pub fn base64(media_type: impl Into<String>, data: impl Into<String>) -> Self {
        Self::Base64 {
            media_type: media_type.into(),
            data: data.into(),
        }
    }

    pub(crate) fn validate(&self) -> Result<()> {
        match self {
            Self::Url { url } if url.trim().is_empty() => Err(MemindError::invalid_request(
                "Source::Url requires a non-blank URL",
            )),
            Self::Base64 { media_type, .. } if media_type.trim().is_empty() => Err(
                MemindError::invalid_request("Source::Base64 requires a non-blank media_type"),
            ),
            Self::Base64 { data, .. } if data.is_empty() => Err(MemindError::invalid_request(
                "Source::Base64 requires non-empty data",
            )),
            _ => Ok(()),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum RawContent {
    Conversation {
        messages: Vec<Message>,
    },
    Object {
        content_type: String,
        fields: serde_json::Map<String, serde_json::Value>,
    },
}

impl RawContent {
    pub fn conversation(messages: Vec<Message>) -> Result<Self> {
        let raw = Self::Conversation { messages };
        raw.validate()?;
        Ok(raw)
    }

    pub fn object(
        content_type: impl Into<String>,
        fields: serde_json::Map<String, serde_json::Value>,
    ) -> Result<Self> {
        let content_type = content_type.into().trim().to_string();
        if content_type.is_empty() {
            return Err(MemindError::invalid_request(
                "RawContent::object requires a non-blank content_type",
            ));
        }
        if content_type == "conversation" {
            return Err(MemindError::invalid_request(
                "RawContent::object cannot use content_type \"conversation\"; use RawContent::conversation",
            ));
        }
        if fields.contains_key("type") {
            return Err(MemindError::invalid_request(
                "RawContent::object fields must not contain \"type\"",
            ));
        }
        Ok(Self::Object {
            content_type,
            fields,
        })
    }

    pub(crate) fn validate(&self) -> Result<()> {
        match self {
            Self::Conversation { messages } => {
                if messages.is_empty() {
                    return Err(MemindError::invalid_request(
                        "conversation raw content requires at least one message",
                    ));
                }
                for message in messages {
                    message.validate()?;
                }
                Ok(())
            }
            Self::Object {
                content_type,
                fields,
            } => {
                if content_type.trim().is_empty() {
                    return Err(MemindError::invalid_request(
                        "RawContent::object requires a non-blank content_type",
                    ));
                }
                if content_type == "conversation" {
                    return Err(MemindError::invalid_request(
                        "RawContent::object cannot use content_type \"conversation\"; use RawContent::conversation",
                    ));
                }
                if fields.contains_key("type") {
                    return Err(MemindError::invalid_request(
                        "RawContent::object fields must not contain \"type\"",
                    ));
                }
                Ok(())
            }
        }
    }
}

impl Serialize for RawContent {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match self {
            Self::Conversation { messages } => {
                let mut map = serializer.serialize_map(Some(2))?;
                map.serialize_entry("type", "conversation")?;
                map.serialize_entry("messages", messages)?;
                map.end()
            }
            Self::Object {
                content_type,
                fields,
            } => {
                let mut map = serializer.serialize_map(Some(fields.len() + 1))?;
                map.serialize_entry("type", content_type)?;
                for (key, value) in fields {
                    map.serialize_entry(key, value)?;
                }
                map.end()
            }
        }
    }
}

impl<'de> Deserialize<'de> for RawContent {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = serde_json::Value::deserialize(deserializer)?;
        let mut object = match value {
            serde_json::Value::Object(object) => object,
            _ => return Err(de::Error::custom("raw content must be a JSON object")),
        };

        let content_type = object
            .remove("type")
            .and_then(|value| value.as_str().map(ToOwned::to_owned))
            .ok_or_else(|| de::Error::custom("raw content requires string field \"type\""))?;

        if content_type == "conversation" {
            let messages_value = object
                .remove("messages")
                .ok_or_else(|| de::Error::custom("conversation raw content requires messages"))?;
            let messages =
                Vec::<Message>::deserialize(messages_value).map_err(de::Error::custom)?;
            return Ok(Self::Conversation { messages });
        }

        Ok(Self::Object {
            content_type,
            fields: object,
        })
    }
}

File: memind-clients/typescript/src/types/message.ts
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { Role } from './common.js'

export type UrlSource = {
  type: 'url'
  url: string
}

export type Base64Source = {
  type: 'base64'
  media_type: string
  data: string
}

export type Source = UrlSource | Base64Source

export type TextBlock = { type: 'text'; text: string }
export type ImageBlock = { type: 'image'; source: Source }
export type AudioBlock = { type: 'audio'; source: Source }
export type VideoBlock = { type: 'video'; source: Source }

export type ContentBlock = TextBlock | ImageBlock | AudioBlock | VideoBlock

export type MessageValue = {
  role: Role
  content: ContentBlock[]
  timestamp?: string
  userName?: string
  sourceClient?: string
}

export type MessageOptions = {
  timestamp?: string
}

export const Message = {
  user(text: string, options?: MessageOptions): MessageValue {
    return {
      role: Role.USER,
      content: [{ type: 'text', text }],
      ...options,
    }
  },
  assistant(text: string, options?: MessageOptions): MessageValue {
    return {
      role: Role.ASSISTANT,
      content: [{ type: 'text', text }],
      ...options,
    }
  },
} as const

export type JsonPrimitive = string | number | boolean | null
export type JsonValue = JsonPrimitive | JsonValue[] | { [key: string]: JsonValue }

export type ConversationContent = {
  type: 'conversation'
  messages: MessageValue[]
}

export type JsonObjectRawContent = {
  type: string
  [key: string]: JsonValue
}

export type AgentTimelineEvent = {
  eventId?: string
  seq?: number
  kind?: string
  occurredAt?: string
  text?: string
  toolName?: string
  input?: JsonValue
  output?: JsonValue
  status?: string
  durationMs?: number
  path?: string
  operation?: string
  command?: string
  exitCode?: number
  metadata?: Record<string, JsonValue>
}

export type AgentTimelineContent = {
  type: 'agent_timeline'
  sourceClient?: string
  sourceVersion?: string
  sessionId: string
  agentTurnId: string
  timelineId: string
  events: AgentTimelineEvent[]
  project?: Record<string, JsonValue>
  metadata?: Record<string, JsonValue>
}

export type RawContentValue = ConversationContent | AgentTimelineContent | JsonObjectRawContent

export const RawContent = {
  conversation(messages: MessageValue[]): ConversationContent {
    if (!Array.isArray(messages) || messages.length === 0) {
      throw new Error('RawContent.conversation requires a non-empty messages array')
    }
    for (const message of messages) {
      assertMessage(message)
    }
    return { type: 'conversation', messages }
  },

  map(type: string, fields?: Record<string, JsonValue>): JsonObjectRawContent {
    if (!type || !type.trim()) {
      throw new Error('RawContent.map requires a non-blank type')
    }
    if (type === 'conversation') {
      throw new Error(
        'RawContent.map cannot use type "conversation" - use RawContent.conversation() instead',
      )
    }
    if (fields && 'type' in fields) {
      throw new Error('RawContent.map fields must not contain a "type" key')
    }
    if (fields) {
      assertJsonValue(fields, 'fields')
    }
    return { type, ...fields }
  },
} as const

export function assertRawContent(rawContent: RawContentValue): void {
  if (!rawContent || typeof rawContent !== 'object' || Array.isArray(rawContent)) {
    throw new Error('rawContent must be an object')
  }
  if (typeof rawContent.type !== 'string' || !rawContent.type.trim()) {
    throw new Error('rawContent.type must be a non-blank string')
  }
  if (rawContent.type === 'conversation') {
    const messages = (rawContent as { messages?: unknown }).messages
    if (!Array.isArray(messages) || messages.length === 0) {
      throw new Error('conversation rawContent requires a non-empty messages array')
    }
    for (const message of messages) {
      assertMessage(message as MessageValue)
    }
    return
  }
  assertJsonValue(rawContent, 'rawContent')
}

export function assertMessage(message: MessageValue): void {
  if (!message || typeof message !== 'object') {
    throw new Error('Message must be an object')
  }
  if (message.role !== Role.USER && message.role !== Role.ASSISTANT) {
    throw new Error('Message role must be USER or ASSISTANT')
  }
  if (!Array.isArray(message.content) || message.content.length === 0) {
    throw new Error('Message content must be a non-empty array')
  }
  for (const block of message.content) {
    assertContentBlock(block)
  }
}

function assertContentBlock(block: ContentBlock): void {
  if (!block || typeof block !== 'object') {
    throw new Error('Content block must be an object')
  }
  if (block.type === 'text') {
    if (typeof block.text !== 'string') {
      throw new Error('Text content block requires string text')
    }
    return
  }
  if (block.type === 'image' || block.type === 'audio' || block.type === 'video') {
    assertSource(block.source)
    return
  }
  throw new Error(`Unsupported content block type: ${(block as { type?: unknown }).type}`)
}

function assertSource(source: Source): void {
  if (!source || typeof source !== 'object') {
    throw new Error('Media content block requires a source object')
  }
  if (source.type === 'url') {
    if (typeof source.url !== 'string' || !source.url.trim()) {
      throw new Error('URL source requires a non-blank url')
    }
    return
  }
  if (source.type === 'base64') {
    if (typeof source.media_type !== 'string' || !source.media_type.trim()) {
      throw new Error('Base64 source requires a non-blank media_type')
    }
    if (typeof source.data !== 'string' || !source.data) {
      throw new Error('Base64 source requires data')
    }
    return
  }
  throw new Error(`Unsupported source type: ${(source as { type?: unknown }).type}`)
}

function assertJsonValue(value: unknown, path: string, seen = new WeakSet<object>()): void {
  if (
    value === null ||
    typeof value === 'string' ||
    typeof value === 'boolean' ||
    (typeof value === 'number' && Number.isFinite(value))
  ) {
    return
  }
  if (
    typeof value !== 'object' ||
    value === undefined ||
    typeof value === 'function' ||
    typeof value === 'symbol' ||
    typeof value === 'bigint'
  ) {
    throw new Error(`RawContent.map field "${path}" must be JSON-serializable`)
  }
  if (seen.has(value)) {
    throw new Error(`RawContent.map field "${path}" must not be cyclic`)
  }
  seen.add(value)
  try {
    if (Array.isArray(value)) {
      value.forEach((item, index) => assertJsonValue(item, `${path}[${index}]`, seen))
      return
    }
    for (const [key, item] of Object.entries(value)) {
      assertJsonValue(item, `${path}.${key}`, seen)
    }
  } finally {
    seen.delete(value)
  }
}
